Add support for spaces that are escaped with a backslash

J4yD4n opened the pull request:
Since buf is being split by spaces, the while loop ends up checking for '\ '. If it is found, it will change the backslash in to a space and then append the next word. This is repeated until it runs out of escaped spaces and then the end result string is added to the groups array as normal.

xdesai commented:
Thanks for the PR!
We're tracking this internally right now. No timeline when we'll be able to pull it in just yet because we need to have automated testing around this and validate it on the different operating systems we support. I'll make sure to keep you posted on here when we know more.

vbscott commented:
Hello,
The PR looks good but if it is possible could you change your implementation to use strlcpy and strncat instead of strcpy and strcat.  If possible we also require tests for any new code that is written.

J4yD4n commented:
I've changed the code to use strlcpy and strncat instead of strcpy and strcat. I've also modified the test scripts and config files to be able to test the code as well as created groups-1.t to describe the test process and the returned result.

vbscott commented:
Hi, I wanted to let you know I made a commit to your pull request to add your test to the Makfile.am in the tests directory and merged with a conflict in that Makefile.

J4yD4n commented on lib/util.c:
```diff
@@ -77,6 +77,15 @@ duo_common_ini_handler(struct duo_config *cfg, const char *section,
                 free(buf);
                 return (0);
             }
+            while (p[strlen(p) - 1] == '\\') {
```
I'm not certain how strtok would respond to some calls being inside of a function and some outside. Even if it does work, it sounds like it would make the code more confusing. I added in a comment so it is a little more obvious what the while loop is doing.

J4yD4n commented on lib/util.c:
```diff
@@ -51,8 +51,8 @@ int
 {
-    char *buf, *p;
-    int int_val;
+    char *buf, *p, *q, *r;
```
I renamed the variables p, q, and r so they represent what they do.

J4yD4n commented on lib/util.c:
```diff
@@ -77,6 +77,15 @@ duo_common_ini_handler(struct duo_config *cfg, const char *section,
+            while (p[strlen(p) - 1] == '\\') {
+                p[strlen(p) - 1] = ' ';
+                q = strtok(NULL, " ");
+		new_length = strlen(p) + strlen(q) + 1;
```
I was bouncing between Notepad++ and vim. I fixed the indenting and checked for any other lines that were indented with tabs instead of spaces. They should all be spaces like the rest of the code now.

J4yD4n commented on tests/confs/mockduo_space_users.conf:
```diff
@@ -0,0 +1,6 @@
+skey = yWHSMhWucAcp7qvuH3HWTaSaKABs8Gaddiv1NIRo
+host = localhost:4443
+cafile = certs/mockduo-ca.pem
+groups = users\ with\ spaces
```
I added in tests for the other groups you mentioned and changed "no_space_here" to "no_spaces\here" so it also tests backslashes in the group name. I also removed the second comma so it will test space delimeter group names at the same time.

J4yD4n commented on tests/confs/mockduo_space_users.conf:
```diff
@@ -0,0 +1,6 @@
+skey = yWHSMhWucAcp7qvuH3HWTaSaKABs8Gaddiv1NIRo
+host = localhost:4443
+cafile = certs/mockduo-ca.pem
+groups = users\ with\ spaces
```
I added in testing backslashes in the group name (see above suggestion comment).
I'm not certain how to work with Unity so hopefully one of your developers can convert the test case. Until then, I updated the groups-1.t with the new test cases.
